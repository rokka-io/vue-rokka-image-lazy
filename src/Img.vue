<template>
  <rokka-image-img
    :org="org"
    :hash="hash"
    :filename="filename"
    :stack="stack"
    :format="format"
    :alt="alt"
    :key="hash"
    srcset-attribute="data-srcset"
    src-attribute="data-src"
    src-additional-attribute="src"
    :src-additional="srcAdditionalComputed"
    :postfix="postfix"
    :options="options"
  />
</template>

<script>
import { RokkaImageImg, buildRokkaUrl } from 'vue-rokka-image';
import lozad from 'lozad';

class ImageCache {
  constructor({ max }) {
    this.options = {
      max: max || 100,
    };
    this._caches = [];
  }

  has(key) {
    return this._caches.indexOf(key) > -1;
  }

  add(key) {
    if (this.has(key)) return;
    this._caches.push(key);
    if (this._caches.length > this.options.max) {
      this.free();
    }
  }

  free() {
    this._caches.shift();
  }
}

export default {
  name: 'RokkaImageImgLazy',
  components: {
    RokkaImageImg
  },
  props: {
    ...RokkaImageImg.props,
    srcAttribute: {
      type: String,
      default: 'data-src',
    },
    srcsetAttribute: {
      type: String,
      default: 'data-srcset',
    },
    postfix: {
      type: [Object, Array],
      default: () => ['1x', '2x'],
    },
    options: {
      type: [Object, Array],
      default: () => [{ dpr: '1' }, { dpr: '2' }],
    },
    loading: {
      type: String,
      default: null,
    },
  },
  created() {
    if (!window._imageCache) {
      window._imageCache = new ImageCache({ max: 200 });
    }
  },
  computed: {
    srcAdditionalComputed() {
      if (!this.loading || this.isCached) {
        return "data:image/gif;base64,R0lGODlhAQABAAD/ACwAAAAAAQABAAACADs=";
      }
      window._imageCache.add(this.rokkaRenderUrl);
      return this.loading;
    },

    isCached() {
      return window._imageCache.has(this.rokkaRenderUrl)
    },

    rokkaRenderUrl() {
      return buildRokkaUrl({
        org: this.org,
        hash: this.hash,
        stack: this.stack,
        format: this.format,
        filename: this.filename,
        options: Array.isArray(this.options) ? this.options[0] : this.options,
        variables: Array.isArray(this.variables) ?this.variables[0]: this.variables,
      });

    }
  },
  mounted() {
    // We initialize Lozad.js on the root
    // element of our component.
    if (!window._lozadObserver) {
      window._lozadObserver = lozad('.rokka--attr-data-src', {
        rootMargin: '200px',
      });
    }
    window._lozadObserver.observe();
  },
};
</script>

<style scoped></style>
