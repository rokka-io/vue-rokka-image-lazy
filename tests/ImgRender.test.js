import { mount } from '@vue/test-utils'
import Img from './ImgRender.vue'

let wrapper = mount(Img)

describe('RokkaImg rendering', () => {
  test('renders correctly', () => {
    expect(wrapper.element).toMatchSnapshot()
  })
})
