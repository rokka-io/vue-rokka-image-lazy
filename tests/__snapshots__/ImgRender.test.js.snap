// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RokkaImg rendering renders correctly 1`] = `
<img
  class="rokka--attr-data-src"
  data-cache="https://gallery-demo.rokka.io/dynamic/resize-width-300/o-autoformat-1/7ac62a/image.jpg"
  data-src="https://gallery-demo.rokka.io/dynamic/resize-width-300/o-autoformat-1/7ac62a/image.jpg"
  data-srcset="https://gallery-demo.rokka.io/dynamic/resize-width-300/o-autoformat-1/7ac62a/image.jpg 1x, https://gallery-demo.rokka.io/dynamic/resize-width-300/o-autoformat-1-dpr-2/7ac62a/image.jpg 2x"
  src="https://rokka.io/gallery/assets/loader.svg"
/>
`;
